File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.java
class Solution {
    public int maxProfit(int[] prices, int fee) {
        int n = prices.length;
        int[] nxt = new int[2];
        int[] cur = new int[2];
        
        for (int i = n - 1; i >= 0; i--){
            for (int b = 0; b < 2; b++){
                if (b == 0){
                    cur[b] = Math.max(nxt[0], - prices[i] + nxt[1]);
                }
                else{
                    cur[b] = Math.max(nxt[1], prices[i] - fee + nxt[0]);
                }
            }
            nxt = cur;
        }
        return nxt[0];
    }
}

File: 0714-best-time-to-buy-and-sell-stock-with-transaction-fee/0714-best-time-to-buy-and-sell-stock-with-transaction-fee.py
class Solution(object):
    def maxProfit(self, prices, fee):
        """
        :type prices: List[int]
        :type fee: int
        :rtype: int
        """
        n = len(prices)
        dp = [[-1] * (2) for _ in range(n)]
        def solve(ind, b):
            if ind == n:
                return 0
            if dp[ind][b] != -1:
                return dp[ind][b]
            profit = 0
            if not b:
                profit = max(solve(ind + 1, 0), - prices[ind] + solve(ind + 1, 1))
            else:
                profit = max(solve(ind + 1, 1), prices[ind] - fee + solve(ind + 1, 0))
            dp[ind][b] = profit
            return profit
        return solve(0, 0)